File: services/auto_complete/models/Trie.java
import java.util.HashMap;
import java.util.ArrayList;
import java.util.Map;

public class Trie {

    private final static int TOP_POPULAR_SIZE = 5;

    public static class Node {
        private HashMap<Character, Node> children;
        private ArrayList<String> topPopular;

        public Node() {
            this.children = new HashMap<Character, Node>(); 
            this.topPopular = new ArrayList<String>();
        }

        public HashMap<Character, Node> getChildren() {
            return this.children;
        }

        public void addChild(Character ch, Node node) {
            this.children.put(ch, node);
        }

        public ArrayList<String> getTopPopular() {
            return this.topPopular;
        }

        public void addTopPopular(String word) {
            this.topPopular.add(word);
        }

        @Override
        public String toString() {
            return "Children = " + this.children.size()+ "; "
                + "Top popular = [ " + this.topPopular.toString() + "]\n";
        }
    }

    public static class TrieBuilder {
        private Node root;

        public TrieBuilder() {
            this.root = new Node();
        }

        public Node getRoot() {
            return this.root;
        }

        private void view(Node pa, Node u) {
            System.out.println(u);
            for (Map.Entry<Character, Node> e: u.getChildren().entrySet()) {
                Node v = (Node) e.getValue();
                if (v==pa) {
                    continue;
                }

                this.view(u,v);
            }
        }

        public void addWord(String word) {
            word = word.toLowerCase();
            Node curr = this.root;
            for (char ch: word.toCharArray()) {
                if (curr.getChildren().get(ch) != null) {
                    curr = curr.getChildren().get(ch);
                } else {
                    Node n_curr = new Node(); 
                    curr.addChild(ch, n_curr);
                    curr = n_curr;
                }

                if (curr.getTopPopular().size() < TOP_POPULAR_SIZE) {
                    curr.addTopPopular(word);
                }
            }
        }

        public ArrayList<String> getTopPopular(String prefix) {
            prefix = prefix.toLowerCase();
            Node curr = this.root;

            for (char ch: prefix.toCharArray()) {
                if (curr.getChildren().get(ch) == null) {
                    return new ArrayList<String>();
                }

                curr = curr.getChildren().get(ch);
                //System.out.println(curr);
            }
            
            return curr.getTopPopular();
        }
    }

    public static void main(String[] args) {
    }
}

File: services/auto_complete/models/Trie.py
import os

TOP_POPULAR_SIZE = int(os.getenv("TOP_POPULAR_SIZE"))

class Node:
    def __init__(self):
        #["char" : Node()]
        self.children = dict() 
        self.top_popular = set()
    
    def __str__(self):
        return f'Children: {self.children}; top_pick: {self.top_popular}'
    
class Trie:
    def __init__(self):
        self._root = Node() 
    
    @property
    def root(self):
        return self._root
    
    def view(self, pa, u):
        print(u)
        for v in u.children.values():
            if v==pa:
                continue 
            self.view(u,v)
    
    def add_word(self, word):
        word = word.lower() 
        curr = self._root 
        for ch in word:
            if ch in curr.children:
                curr = curr.children.get(ch) 
            else:
                n_curr = Node() 
                curr.children[ch] = n_curr
                curr = n_curr 
            
            if len(curr.top_popular) < TOP_POPULAR_SIZE:
                curr.top_popular.add(word)
    
    def get_top_popular(self, prefix):
        prefix = prefix.lower() 
        curr = self._root
        
        for ch in prefix:
            if ch not in curr.children:
                return set() 
            
            curr = curr.children[ch]
        
        return list(curr.top_popular)
    
if __name__ == "__main__":
    trie = Trie() 
    trie.add_word("abc")
    trie.add_word("aac")
    trie.add_word("aab")
    trie.add_word("abd")
    trie.add_word("acd")
    trie.add_word("ace")
    trie.add_word("acc")    
    trie.view(trie.root, trie.root)
    print(trie.get_top_popular("ab"))
    